File: samples/csharp/hosted-agents/bring-your-own/responses/browser-automation/src/browser-automation-csharp-byo-sample-foundry/utils/Skills.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Reflection;

namespace BrowserAutomation;

/// <summary>
/// Skills manager — loads embedded markdown skill files for guided workflows.
/// </summary>
public static class Skills
{
    private static readonly Assembly Assembly = typeof(Skills).Assembly;

    /// <summary>Load a skill markdown file by name.</summary>
    public static string? LoadSkill(string name)
    {
        // Sanitize name
        var safeName = new string(name.Where(c => char.IsLetterOrDigit(c) || c == '-' || c == '_').ToArray());
        var resourceName = Assembly.GetManifestResourceNames()
            .FirstOrDefault(r => r.EndsWith($".{safeName}.md", StringComparison.OrdinalIgnoreCase));

        if (resourceName == null) return null;

        using var stream = Assembly.GetManifestResourceStream(resourceName);
        if (stream == null) return null;
        using var reader = new StreamReader(stream);
        return reader.ReadToEnd();
    }

    /// <summary>List available skill names.</summary>
    public static List<string> ListSkills()
    {
        return Assembly.GetManifestResourceNames()
            .Where(r => r.EndsWith(".md", StringComparison.OrdinalIgnoreCase))
            .Select(r =>
            {
                // Resource names are like "browser_automation.skills.form-filler.md"
                var parts = r.Split('.');
                return parts.Length >= 2 ? parts[^2] : r;
            })
            .ToList();
    }
}

File: samples/python/hosted-agents/bring-your-own/responses/browser-automation/src/bat-python-byo/utils/skills.py
# Copyright (c) Microsoft. All rights reserved.

"""Skills manager — loads markdown skill files for guided workflows."""

from __future__ import annotations

import os
import re

_SKILLS_DIR = os.path.join(os.path.dirname(__file__), "..", "skills")


def load_skill(name: str) -> dict:
    """Load a skill markdown file by name."""
    safe_name = re.sub(r"[^a-zA-Z0-9_-]", "", name)
    path = os.path.join(_SKILLS_DIR, f"{safe_name}.md")
    if not os.path.isfile(path):
        available = list_skills()
        return {"error": f"Skill '{name}' not found. Available: {available}"}
    with open(path, "r", encoding="utf-8") as f:
        return {"skill": name, "instructions": f.read()}


def list_skills() -> list[str]:
    """List available skill names."""
    if not os.path.isdir(_SKILLS_DIR):
        return []
    return [f[:-3] for f in os.listdir(_SKILLS_DIR) if f.endswith(".md")]
